src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -408,9 +408,47 @@ private class Emit(
           xvout := xmout.mux(defaultValue(typ), xvout)
         ), xmout, xvout)
 
-      case x@ApplyAggOp(a, op, args) =>
-        val agg = AggOp.get(op, x.inputType, args.map(_.typ))
-        present(emitAgg(a)(agg.seqOp(aggregator, _, _)))
+      case ArrayFor(a, valueName, body) =>
+        val tarray = coerce[TArray](a.typ)
+        val eti = typeToTypeInfo(tarray.elementType)
+        val xmv = mb.newField[Boolean]()
+        val xvv = coerce[Any](mb.newField(valueName)(eti))
+        val bodyenv = env.bind(
+          valueName -> (eti, xmv.load(), xvv.load()))
+        val codeB = emit(body, env = bodyenv)
+        val aBase = emitArrayIterator(a)
+        val cont = { (m: Code[Boolean], v: Code[_]) =>
+          Code(
+            xmv := m,
+            xvv := v,
+            codeB.setup)
+        }
+
+        val processAElts = aBase.arrayEmitter(cont)
+        val ma = processAElts.m.getOrElse(const(false))
+        EmitTriplet(
+          Code(
+            processAElts.setup,
+            ma.mux(
+              Code._empty,
+              Code(aBase.calcLength, processAElts.addElements))),
+          const(false),
+          Code._empty)
+
+      case SeqOp(a, i, agg) =>
+        EmitTriplet(
+          emitAgg(a, env)(agg.seqOp(aggregator(i), _, _)),
+          const(false),
+          Code._empty)
```

Comment:
yeah I would kind of think the emitAgg should be down here ... but now I'm not sure.